### DISCUSSION TITLE
Add a color picker/color checker to define conditional format measures

### DISCUSSION THREAD

**Discussion by bernatagulloesbrina** [TabularEditor/TabularEditor3#1164] [Ideas]
### Is your feature request related to a problem? Please describe

not really a problem but a hassle. 
Defining conditional formatting in power bi desktop is quite of a pain. 
Formatting by Field Value is better, but then you need to figure out the HEX codes, and if you look at the code you may not identify what color is each code. 

### Describe the solution you'd like

it would be nice to be able to right click (or something) while typing DAX and have the option to "inser HEX code" 
Selecting a hex code should equally show up a the color selected in a color picker, in case it needs changing it. 

as an option it could include the icon names in case we are defining a measure for conditional formatting with icons. (but that's much lower priority for me)

### Goals

1. Frictionless definition of conditional format measures
2. Frictionless maintenance of condtional format measures

**Comment by mlonsk** [comment]
Hi @bernatagulloesbrina 
Thank you for the suggestion, I can see how this would improve the ability two write HEX codes. I have added it to our backlog, but cannot give a timeline right now.
My initial thought is that it quite a niche use case, but if I am wrong please help make sure the idea is upvoted 😁

**Comment by otykier** [comment]
Personally, I really don't like the idea of having measures in my model that return hex codes (and don't get me started on SVG measures). So I agree with Morten that this is a very niche case, and I don't think it's something we will prioritise, to be honest.

That being said, there's a very straightforward solution using C# scripting, thanks to the WinForms [ColorDialog class](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.colordialog?view=windowsdesktop-7.0&redirectedfrom=MSDN).

1. Create a C# script with the following code:

```csharp
using System.Windows.Forms;
var colorDialog = new ColorDialog();

// If user cancelled the dialog, do nothing:
if (colorDialog.ShowDialog() != DialogResult.OK) return;

var c = colorDialog.Color;
var hex = "\"#" + c.R.ToString("X2") + c.G.ToString("X2") + c.B.ToString("X2") + "\"";

// Copy the hex string to clipboard (useful in TE2):
// Clipboard.SetText(hex);

// Alternatively, simulate key presses corresponding to the hex string
// (useful when saved as a macro with a custom shortcut in TE3):
SendKeys.Send(hex);
```

2. Save as a macro.
3. Assign a custom shortcut to the macro under **Tools > Preferences > Keyboard**.

Result:

![GetColorHex](https://github.com/TabularEditor/TabularEditor3/assets/8976200/c0b79dbc-af7e-440e-a408-ab21ced7ba30)

**Reply by otykier** [reply]
`SendKeys.Send(...)` is not ideal, as it literally just sends the string as keystrokes to the application in focus - so the script won't work if you're not currently editing a DAX expression. Also, there is no supported way for a script to know the text currently selected in the code editor, so it's difficult to pre-populate the color dialog with the currently selected hex code...

However, we do have some ideas to improve the scripting API in both TE2 and TE3. One important improvement in this area, is that you will have access to the currently focused document, including the text content and the current selection/position of the cursor - both for read and write purposes. This will effectively let you create C# scripts that lets you interact with the current code document in a much more elegant way - for example, to add small snippets of code, swap out placeholder values, use DAX templates, etc...

**Reply by bernatagulloesbrina** [reply]
wow! the script is next level as usual! will look in detail
would it be possible to select some text and make the pop-up show with the selected color? 
some macros to work on selected text would be very useful when authoring code

**Reply by otykier** [reply]
@bernatagulloesbrina unfortunately, there is no supported way to get the text under the cursor/current selection, but we will make sure that becomes possible once we ship the scripting API improvements that I mentioned above.

**Reply by otykier** [reply]
One possible workaround, is that the user copies the hex code into the clipboard (Ctrl+C) before running the script. The script can then examine the clipboard. If the clipboard contains a valid hex code, that will be used as the initial color of the popup:

```csharp
using System.Windows.Forms;
using System.Drawing;

ScriptHelper.WaitFormVisible = false;

var colorDialog = new ColorDialog() { AllowFullOpen = true, FullOpen = true };

try
{
    // Try to see if the clipboard contains a hex representation of a color. Any of the following
    // formats can be used:
    // "#AFD056"
    // #AFD056
    // AFD056
    var clipboardText = Clipboard.GetText().Replace("\"", "");
    if(clipboardText.Length == 6) clipboardText = "#" + clipboardText;
    if(clipboardText.Length == 7 && clipboardText.StartsWith("#"))
        colorDialog.Color = ColorTranslator.FromHtml(clipboardText);    
}
catch
{
    // The clipboard did not contain a valid color, so do nothing.
}

// If user cancelled the dialog, do nothing:
if (colorDialog.ShowDialog() != DialogResult.OK) return;

var c = colorDialog.Color;
var hex = "\"#" + c.R.ToString("X2") + c.G.ToString("X2") + c.B.ToString("X2") + "\"";

// Copy the hex string to clipboard (useful in TE2):
// Clipboard.SetText(hex);

// Alternatively, simulate key presses corresponding to the hex string
// (useful when saved as a macro with a custom shortcut in TE3):
SendKeys.Send(hex);
```

![GetColorHex2](https://github.com/TabularEditor/TabularEditor3/assets/8976200/0cd8564d-2170-41c1-b830-1f08a95d54bb)

**Reply by otykier** [reply]
The `ColorDialog` class also has a [CustomColors](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.colordialog.customcolors?view=windowsdesktop-7.0) property, which is an array of integer values representing the (up to) 16 custom colors at the bottom of the dialog. If you like, you can save this array as an annotation on the model whenefter the ColorDialog is closed. Before showing the ColorDialog, you can then load the same annotation into the CustomColors property. That way, each model can have up to 16 custom colors to make it easy to find and reuse colors at the model level.
